controllers/keystoneapi_controller.go

```diff
@@ -197,6 +199,7 @@ func (r *KeystoneAPIReconciler) reconcileDelete(ctx context.Context, instance *k
 
 	// Service is deleted so remove the finalizer.
 	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
```

Comment:
in here before we remove the finalizer on the keystoneapi instance, we have to remove the finalizer on the db so that it can get deleted. so maybe best to add a Delete func to the db module which right now only removes the finalizer, but could be extended with other cleanup tasks if needed. so something like

~~~
db := database.NewDatabase(
	instance.Name,
	instance.Spec.DatabaseUser,
	instance.Spec.Secret,
	map[string]string{
		"dbName": instance.Spec.DatabaseInstance,
	},
)
// delete the DB
ctrlResult, err := db.Delete(
	ctx,
	helper,
)
~~~

Instead of the database.NewDatabase(..) we should add a func GetDBWithName(...) to get the db if it exist, like it is there fore service [1].

[1] https://github.com/openstack-k8s-operators/keystone-operator/blob/master/pkg/external/funcs.go#L236

Reply:
**ok I'm adding a Delete func in lib-common**

Reply:
wip

Reply:
@stuggi I want to put a getDB if exists function in : keystone_operator_pkomarov/pkg/external/funcs.go 
then use it in keystoneapi_controller.go to check if db exists then remove the db finalizer.
 
Not sure about how to implement the getDb function though something like : ?  
```
//TODO add a get GetDBWithName here like the above ^^^
func GetDBWithName(
	ctx context.Context,
	h *helper.Helper,
)  {
	//use this perhaps ? 
	db := database.GetDatabase()
	return d.database

	}
	// or this ?
	db := database.NewDatabase(
		instance.Name,
		instance.Spec.DatabaseUser,
		instance.Spec.Secret,
		map[string]string{
			"dbName": instance.Spec.DatabaseInstance,
		}
	)

	if db == nil {
		if k8s_errors.IsNotFound(db) {
		return db, err}
}
```

Reply:
> @stuggi I want to put a getDB if exists function in : keystone_operator_pkomarov/pkg/external/funcs.go then use it in keystoneapi_controller.go to check if db exists then remove the db finalizer.

I think it should be a func in the db module of lib-common, so that it can be used by any operator. 


> 
> Not sure about how to implement the getDb function though something like : ?

you should have a parameter with the name and namespace, like [1] and then the get.

[1] https://github.com/openstack-k8s-operators/keystone-operator/blob/master/api/v1beta1/funcs.go#L235-L239

> 
> ```
> //TODO add a get GetDBWithName here like the above ^^^
> func GetDBWithName(
> 	ctx context.Context,
> 	h *helper.Helper,
> )  {
> 	//use this perhaps ? 
> 	db := database.GetDatabase()
> 	return d.database
> 
> 	}
> 	// or this ?
> 	db := database.NewDatabase(
> 		instance.Name,
> 		instance.Spec.DatabaseUser,
> 		instance.Spec.Secret,
> 		map[string]string{
> 			"dbName": instance.Spec.DatabaseInstance,
> 		}
> 	)
> 
> 	if db == nil {
> 		if k8s_errors.IsNotFound(db) {
> 		return db, err}
> }
> ```

Reply:
would this work (code below) ? 
in pkg/external/funcs.go:
My concern is that's it's identical to the getDBWithName function (just with added name and namespace params)
can we instead modify the getDBWithName func to include name and namespace params , and hange the function usage in lib-common ? 

`
func (d *Database) GetServiceDbByName(
	ctx context.Context,
	h *helper.Helper,
	name string,
	namespace string,
) error {

	db := &mariadbv1.MariaDBDatabase{}
	err := h.GetClient().Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, db)
	if err != nil {
		if k8s_errors.IsNotFound(err) {
			return util.WrapErrorForObject(
				fmt.Sprintf("Failed to get %s database %s ", d.databaseName, h.GetBeforeObject().GetNamespace()),
				h.GetBeforeObject(),
				err,
			)
		}

		return util.WrapErrorForObject(
			fmt.Sprintf("DB error %s %s ", d.databaseName, h.GetBeforeObject().GetNamespace()),
			h.GetBeforeObject(),
			err,
		)
	}

	d.database = db

	return nil
}
`

Reply:
> in lib-common/pkg/external/funcs.go:
it should be in lib-commo/modules/database/funcs.go

We should just use a func, not a method. So remove the receiver and return a pointer to the DB

```
func GetServiceDbByName(
ctx context.Context,
h *helper.Helper,
name string,
namespace string,
) (*mariadbv1.MariaDBDatabase, error) {

db := &mariadbv1.MariaDBDatabase{}
err := h.GetClient().Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, db)
if err != nil {
	if k8s_errors.IsNotFound(err) {
		return nil, util.WrapErrorForObject(
			fmt.Sprintf("Failed to get %s database %s ", d.databaseName, h.GetBeforeObject().GetNamespace()),
			h.GetBeforeObject(),
			err,
		)
	}

	return nil, util.WrapErrorForObject(
		fmt.Sprintf("DB error %s %s ", d.databaseName, h.GetBeforeObject().GetNamespace()),
		h.GetBeforeObject(),
		err,
	)
}
return db, nil
}
```